kv/memberlist/memberlist_client.go

```diff
@@ -568,41 +568,88 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
+	reached, err := m.joinMembersInBatches(ctx, m.cfg.MaxJoinRetries, logger)
+	if err != nil {
+		level.Error(logger).Log("msg", "joining memberlist cluster failed", "err", err, "elapsed_time", time.Since(startTime))
+		return false
+	}
+	level.Info(logger).Log("msg", "joining memberlist cluster succeeded", "reached_nodes", reached, "elapsed_time", time.Since(startTime))
+	return true
+}
 
-	level.Info(m.logger).Log("msg", "joining memberlist cluster", "join_members", strings.Join(m.cfg.JoinMembers, ","))
+// joinMembersInBatches joins m.cfg.JoinMembers 100 at a time. After each batch of 100 it rediscoveres the members.
+// This helps when the list of members is big and by the time we reach the end the originally resolved addresses may be obsolete.
+// joinMembersInBatches returns an error iff it couldn't successfully join any node.
+func (m *KV) joinMembersInBatches(ctx context.Context, numAttempts int, logger log.Logger) (int, error) {
+	const joinNodesBatch = 100
+	var (
+		cfg = backoff.Config{
+			MinBackoff: m.cfg.MinJoinBackoff,
+			MaxBackoff: m.cfg.MaxJoinBackoff,
+			MaxRetries: numAttempts,
+		}
+		boff               = backoff.New(ctx, cfg)
+		lastErr            = error(nil)
+		successfullyJoined = 0
+	)
+
+	for ; boff.Ongoing(); boff.Wait() {
+		var (
+			attemptedNodes = make(map[string]bool) // On each attempt we want to retry all nodes.
+			joinedInBatch  int
+			err            error
+		)
+		for batchHasMore := true; boff.Ongoing() && batchHasMore; successfullyJoined += joinedInBatch {
+			// Rediscover nodes and try to join a subset of them with each batch.
+			// When the list of nodes is large by the time we reach the end of the list some of the
+			// IPs can be unreachable.
+			nodes := m.discoverMembers(ctx, m.cfg.JoinMembers)
+			if len(nodes) == 0 {
+				level.Warn(logger).Log("msg", "joining memberlist cluster: found no nodes to join", "attempts", boff.NumRetries()+1, "max_attempts", numAttempts)
+				lastErr = errors.New("found no nodes to join")
+				break
+			}
 
-	cfg := backoff.Config{
-		MinBackoff: m.cfg.MinJoinBackoff,
-		MaxBackoff: m.cfg.MaxJoinBackoff,
-		MaxRetries: m.cfg.MaxJoinRetries,
+			batchHasMore, joinedInBatch, err = m.joinMembersBatch(nodes, joinNodesBatch, attemptedNodes)
+			if err != nil {
+				lastErr = err
+			}
+		}
+		if successfullyJoined > 0 {
+			// If there are _some_ successful joins, then we can consider the join done.
+			// Mimicking the Join semantics we return an error only when we couldn't join any node at all
+			lastErr = nil
+			break
+		}
 	}
 
-	boff := backoff.New(ctx, cfg)
-	var lastErr error
+	return successfullyJoined, lastErr
+}
 
-	for boff.Ongoing() {
-		// We rejoin all nodes, including those that were joined during "fast-join".
-		// This is harmless and simpler.
-		nodes := m.discoverMembers(ctx, m.cfg.JoinMembers)
+func (m *KV) joinMembersBatch(nodes []string, batchSize int, attemptedNodes map[string]bool) (moreLeft bool, successfullyJoined int, lastErr error) {
```

Comment:
[nit] From a design perspective, passing all `nodes` and `attemptedNodes` looks a bit an anti-pattern to me. I don't think it's responsibility of the function joining a batch of nodes. The caller could build the list of nodes to join, and `joinMembersBatch()` just does it. Also I would take `ctx` in input here, so that you can break the loop as soon as context is cancelled, without waiting for all 100 nodes in the batch to be tried.